### ISSUE TITLE
BadImageFormatException with Span<T> with xUnit.net

### ISSUE THREAD

**Issue by joshfree** [dotnet/coreclr#14097]
Migrated from https://github.com/dotnet/corefx/issues/24135 on behalf of @lextm

> `Span<T>` has been so special, that a strange runtime exception can occur. Reported to xUnit.net already, https://github.com/xunit/xunit/issues/1457 but not sure if the issue is merely caused by xUnit.net. I hope a better exception as well as more meaningful call stack can be thrown if possible.

https://github.com/xunit/xunit/issues/1457#issuecomment-330546584

**Comment by ahsonkhan** [comment]
Updating the corefx issue # listed above (from 2413 to 24135).

**Comment by ahsonkhan** [comment]
When span is boxed (or otherwise used incorrectly), the runtime throws the following exception:
```text
Unhandled Exception: System.BadImageFormatException: An attempt was made to load a program with an incorrect format. (Exception from HRESULT: 0x8007000B)
```

Essentially, this issue is suggesting a more information exception be thrown (see https://github.com/xunit/xunit/issues/1457#issuecomment-330795410).

@RussKeldorph, can you please point me to someone that can help with this?

From: https://github.com/xunit/xunit/issues/1457#issuecomment-330614601
> New compiler will understand the limitations of Span and Span-containing types.
Newer version of Span will not be usable from old compilers.
This scenario will be a compiler error - one way or another, once all pieces are in place.
Here you are using Span with an old compiler that does not understand the safety implications. This is a point-int-time problem. Hopefully very brief.

@VSadov, is this issue still relevant given that Span cannot is not usable from old compilers? I see a warning when I use Span with an old compiler:
>  warning CS0618: 'Span<byte>' is obsolete: 'Types with embedded references are not supported in this version of your compiler.'

**Comment by VSadov** [comment]
You should be getting an error, not a mere warning with the most recent version of Span.
 Notice the `true` argument in the `[Obsolete]`

```C#
namespace System
{
    //
    // Type parameters:
    //   T:
    [DebuggerDisplay("{DebuggerDisplay,nq}")]
    [DebuggerTypeProxy(typeof(System.SpanDebugView<>))]
    [DefaultMember("Item")]
    [IsByRefLikeAttribute]
    [IsReadOnlyAttribute]
    [Obsolete("Types with embedded references are not supported in this version of your compiler.", true)]
    public struct Span<T>
    {
        //
        // Parameters:
        //   array:
        public Span(T[] array);
        //
        // Parameters:
        //   pointer:
        //
        //   length:
        [CLSCompliant(false)]
        public Span(void* pointer, int length);
        //
        // Parameters:
        //   array:
        //
        //   start:
        //
        //   length:
        public Span(T[] array, int start, int length);
. . . 
```

**Comment by VSadov** [comment]
I do not think this issue is very relevant now.

**Comment by ahsonkhan** [comment]
> You should be getting an error, not a mere warning with the most recent version of Span.

Yep. Updated to the latest runtime `<RuntimeFrameworkVersion>2.1.0-preview1-25929-01</RuntimeFrameworkVersion>` and I see the messages as errors now.

**Comment by RussKeldorph** [comment]
So can this issue be closed?

**Comment by VSadov** [comment]
I think this one can be closed.  

C# compiler will prevent misuse of ref-like types that could put them on heap or in generics or in other invalid situations. And we have blocked the use of such types from the old compilers. 

There are still ways to get around the compiler restrictions by using `dynamic`, reflection, RefEmit, and so on. Some of these cases seem unlikely and self-inflicted. There is still a possibility that users could get in trouble and not know why. 

Basically, the cases where ref-like ends up being used incorrectly, will still be a JIT error.
If there is a desire to make such errors better than
```
Unhandled Exception: System.BadImageFormatException: An attempt was made to load a program with an incorrect format. (Exception from HRESULT: 0x8007000B)
```

then another issue should be logged - specifically to improve the error

**Comment by ahsonkhan** [comment]
Closing this issue.

> Basically, the cases where ref-like ends up being used incorrectly, will still be a JIT error.
If there is a desire to make such errors better ...
... then another issue should be logged - specifically to improve the error

@lextm, @stephentoub - let me know if this is still a concern and if you want to see better runtime errors.
